src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java

```diff
@@ -0,0 +1,202 @@
+        }
+        if (detectorNamePattern != null) {
+            mustList.add(new WildcardQueryBuilder("name.keyword", detectorNamePattern));
+        }
+        if (indices != null) {
+            mustList.add(new TermQueryBuilder("indices", indices));
+        }
+        if (highCardinality != null) {
+            mustList.add(new TermQueryBuilder("detector_type", highCardinality ? "MULTI_ENTITY" : "SINGLE_ENTITY"));
+        }
+        if (lastUpdateTime != null) {
+            mustList.add(new BoolQueryBuilder().filter(new RangeQueryBuilder("last_update_time").gte(lastUpdateTime)));
+
+        }
+
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.must().addAll(mustList);
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+            .query(boolQueryBuilder)
+            .size(size)
+            .from(startIndex)
+            .sort(sortString, sortOrder);
+
+        SearchRequest searchDetectorRequest = new SearchRequest().source(searchSourceBuilder);
+
+        if (running != null || disabled != null || failed != null) {
+            // TODO: add a listener to trigger when the first response is received, to trigger the profile API call
+            // to fetch the detector state, etc.
+            // Will need AD client to onboard the profile API first.
+        }
+
+        ActionListener<SearchResponse> searchDetectorListener = ActionListener.<SearchResponse>wrap(response -> {
+            StringBuilder sb = new StringBuilder();
+            SearchHit[] hits = response.getHits().getHits();
+            sb.append("AnomalyDetectors=[");
+            for (SearchHit hit : hits) {
+                sb.append("{");
+                sb.append("id=").append(hit.getId()).append(",");
+                sb.append("name=").append(hit.getSourceAsMap().get("name"));
+                sb.append("}");
+            }
+            sb.append("]");
+            sb.append("TotalAnomalyDetectors=").append(response.getHits().getTotalHits().value);
+            listener.onResponse((T) sb.toString());
+        }, e -> { listener.onFailure(e); });
+
+        adClient.searchAnomalyDetectors(searchDetectorRequest, searchDetectorListener);
+    }
+
+    @Override
+    public boolean validate(Map<String, String> parameters) {
+        return true;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    /**
+     * Factory for the {@link SearchAnomalyDetectorsTool}
+     */
+    public static class Factory implements Tool.Factory<SearchAnomalyDetectorsTool> {
+        private Client client;
+
+        private AnomalyDetectionNodeClient adClient;
+
+        private static Factory INSTANCE;
+
+        /** 
+         * Create or return the singleton factory instance
+         */
+        public static Factory getInstance() {
+            if (INSTANCE != null) {
+                return INSTANCE;
+            }
```

Comment:
I know you just copied this from pretty much all the other tools (including one I wrote that may have started the trend) but I think this whole double-checked-locking bit is overkill for this factory method.  But may as well leave it here for consistency!

Reply:
Agree, currently all getInstance() method are invoked in plugin initialization which is single thread. And here is a bug even the assumption is multiple threads will access to this method, which is: the INSTANCE is missing `volatile` decorator and another thread could be using a partial constructed object and causing exception.

Reply:
Which is why I should have reflexively objected to double-checked-locking at the start.  This probably needs a repo-wide fix to just wrap everything in a synchronized method and be done with it. It's not hot-path.